src/puppetlabs/puppetdb/queue.clj

```diff
@@ -30,15 +32,78 @@
     (catch Exception e
       (throw+ {:kind ::parse-error} e "Error parsing command"))))
 
+(defn sanitize-certname
+  "Replace any underscores and filename forbidden characters found in `certname`
+  with dashes."
+  [certname]
+  (let [forbidden-chars (conj constants/filename-forbidden-characters
+                              "_")]
+    (str/replace
+      certname
+      (re-pattern (str "("
+                       (->> forbidden-chars
```

Comment:
Sounds plausible to me, and then it might even be justifiable to try to get it "exactly right", i.e. more like what I was (not) thinking when I used re-quote-replacement in the first place, and that might be as trivial as figuring out if there's an escape for \E within \Q \E (or just reject \E).